### ISSUE TITLE
[IOS] product does not reach the "approved" state (random behaviour)

### ISSUE THREAD

**Issue by jancellor** [j3k0/cordova-plugin-purchase#993]
### System info

```
macOS Mojave 10.14.6 development machine, XCode 11.3.1
Cordova 9.0.0
Device: iPhone iOS 12.4.4 // Android 7.1
Plugin Version: 9.0.1 (I realise 10.0.1 is out)
```

### Applies to

All the following applies to iOS and development builds (XCode > archive > distribute > development) and/or logged in with a user who has been added as an external tester in the TestFlight section of App Store Connect. The iOS purchase pop-ups either include the text "[Environment: Sandbox]" or something about beta testers not being charged. I have tried this on dev build on my test device with my app store connect user logged in and on three TestFlight users devices. TestFlight is the further I've got - no production releases yet.

### Expected behaviour

After registering and buying the non-consumable product, I expect the purchase to reach the "approved" state.

### Actual behaviour

Instead it goes through "requested" and "initiated" then goes back to the "valid" state. But on a later attempt at the same process, the purchase then goes to "approved". I think it is random whether the purchase gets to approved or not (otherwise it's always on the second attempt but I don't think so).

### Other info

I am calling in sequence at app startup: `register(...),` `when("product").updated(...)`, `error(...)`, `refresh()`. I have a button in the app which calls `order(...)`. I also have a button in the app which calls (again) `refresh()` (though I'm pretty sure refreshing doesn't ever change the behaviour). The first line in the `when("product").updated(...)` call is to log the status change.

The follow may also be relevant. The UI does not quite get updated properly after I call `order(...)`. A div becomes visible saying that the purchase is in progress when the "initiated" state occurs but the bottom half of this div gets cut off. This state-change happens while the the OS password pop-up appears. But the div is only cut off when the purchase is about to fail (I think). Can it be that there is some kind of timing issue? Can the "approved" state change event not be fired if the "initiated" event handler takes too long? I can't see how this would make sense.

I am also getting updates from a product whose ID is the same as that of my _app_ itself. This is very weird. I have never registered a product with that ID. I assume this is not relevant.

### Questions

Any help much appreciated. Any reason to think this happens in dev builds/TestFlight and not production? Anything else useful I can do to figure out what is going on?

**Comment by j3k0** [comment]
Did you solve the issue? Can you try with the latest version of the plugin?

**Comment by jancellor** [comment]
Hi Jean-Christophe. I tried with the latest version of the plugin yesterday and made three purchases which were all successful first time. I can't be sure if the change from v9 to v10 has fixed the issue or if it's just that something else was different or that I got lucky. (The only way I know to test non-consumable purchases more than once is to sign in with a new Apple ID or to create dummy products in App Store Connect so I can't do it quickly.)

Does it sound plausible to you that there was a change from v9 to v10 that got rid of non-deterministic behaviour? Just to reiterate, this was for TestFlight and dev builds, where the "purchase" was free in case that affects anything.

I also released to production yesterday so hopefully I will hear if there are any issues from users.
